fix: improve path collision detection and simplify check

simllll opened the pull request:
### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
- [x] Have you written unit tests?
- [x] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
- [x] This PR is associated with an existing issue?

closes #890 #894

**Test plan**

I added 2 more cases to the route tests:
* two parameters in one route without a folder
* two routes but with a different "prefix"

simllll commented on packages/cli/src/metadataGeneration/metadataGenerator.ts:
```diff
@@ -113,7 +113,7 @@ export class MetadataGenerator {
 
   private checkForPathParamSignatureDuplicates = (controllers: Tsoa.Controller[]) => {
-    const isParamRegExp = new RegExp('{|}|:');
+    const paramRegExp = new RegExp('{(\\w*)}|:(\\w+)', 'g');
```
is there any regex pre defined for parsing the parameters?

WoH replied:
Ideally, for express/koa: Replace `{}` with `:` and feed that into [path-to-regexp](https://www.npmjs.com/package/path-to-regexp)

simllll replied:
I understand, I will push a new rewrite with simplifies and improves the check even more.. and kicks out "false flags" ;)

simllll commented on packages/cli/src/metadataGeneration/metadataGenerator.ts:
```diff
@@ -125,16 +125,12 @@ export class MetadataGenerator {
-          (!tester.paths[i].isParam && subject.paths[i].isParam) ||
-          (!tester.paths[i].isParam && tester.paths[i].path !== subject.paths[i].path)
-        ) {
+        if (tester.paths[i].cntParams !== subject.paths[i].cntParams || tester.paths[i].path !== subject.paths[i].path) {
```
just came into my mind: I think we could probably even more simplify this check, cntParms is not exactly needed anymore, as I compare the path with empty placeholders now.

simllll replied:
simplified it

simllll commented:
I refactored the code, simplified it and improved the collision detection... it should not trigger any "fake" warnings any more, but it warns in two cases now:
1.) if the path is exactly the same (regardless if they contain a parameter or not, but probably most common with parameters ;-))
2.) if the path collides in order, this check only applies to routes with parameters: e.g. 
a. /route-:param1
b. /route-:param2-:param3
Here route b collides with route a. As "param1" could be "a-b".

This one will not collide though:
a. /route-:param1
b. /route-:param2/:param3
as the path is seperated by a "/".

I also changed the output of the warnings.

simllll commented:
This issue (and the unreleased issue #896) is quite a show stopper for our team right now. Anything I can do to bring this into master and publish a new release? :) thanks a lot for all your effort @WoH

WoH commented:
> This issue (and the unreleased issue #896) is quite a show stopper for our team right now. Anything I can do to help bringing this into master and a get new tsoa release? :) thanks a lot for all your effort @WoH 

I'm sorry about that. The code looked good, but I wanted to play around with it for myself to prevent any avoidable issues. LGTM, thanks a lot for the PR!
